File: Pattern-DynamicProgramming/PalindromicSubsequences.js/longestPalindromicSubsequence/recursive.js
// Given a sequence, find the length of its Longest Palindromic Subsequence (LPS). In a palindromic subsequence, elements read the same backward and forward.

// A subsequence is a sequence that can be derived from another sequence by deleting some or no elements without changing the order of the remaining elements.

// Example 1:

// Input: "abdbca"
// Output: 5
// Explanation: LPS is "abdba".

// Example 2:

// Input: = "cddpd"
// Output: 3
// Explanation: LPS is "ddd".

// Example 3:

// Input: = "pqr"
// Output: 1
// Explanation: LPS could be "p", "q" or "r".

const findLPSLength = function (st) {
  function recursive(st, start, end) {
    if (start > end) {
      return 0;
    }

    if (start === end) {
      return 1;
    }

    if (st[start] === st[end]) {
      return 2 + recursive(st, start + 1, end - 1);
    }

    let a = recursive(st, start + 1, end);
    let b = recursive(st, start, end - 1);
    return Math.max(a, b);
  }
  return recursive(st, 0, st.length - 1);
};

// Time complexity is O(2^N). Even though we have three recursive calls we will only ever make 1 or 2. N is the length of the input string

// Space complexity is is O(N)

console.log('Length of LPS ---> ' + findLPSLength('abdbca'));
console.log('Length of LPS ---> ' + findLPSLength('cddpd'));
console.log('Length of LPS ---> ' + findLPSLength('pqr'));

File: Pattern-DynamicProgramming/LongestCommonSubstring/longestRepeatingSubsequence/recursive.py
# Given a sequence, find the length of its longest repeating subsequence (LRS). A repeating subsequence will be the one that appears at least twice in the original sequence and is not overlapping (i.e. none of the corresponding characters in the repeating subsequences have the same index).

# Example 1:

# Input: “t o m o r r o w”
# Output: 2
# Explanation: The longest repeating subsequence is “or” {tomorrow}.

# Example 2:

# Input: “a a b d b c e c”
# Output: 3
# Explanation: The longest repeating subsequence is “a b c” {a a b d b c e c}.

# Example 3:

# Input: “f m f f”
# Output: 2
# Explanation: The longest repeating subsequence is “f f” {f m f f, f m f f}. Please note the second last character is shared in LRS.

def find_LRS_length(str):

    return lrs(str,0,0)

def lrs(str,idx1,idx2):
    if idx1 == len(str) or idx2 == len(str):
        return 0

    if idx1 != idx2 and str[idx1] == str[idx2]:
        return 1 + lrs(str,idx1+1,idx2+1)

    c1 = lrs(str,idx1+1,idx2)
    c2 = lrs(str,idx1,idx2+1)
    return max(c1,c2)

# Time complexity is exponential O(2^n)

# Space complexity is O(N)

def main():
  print(find_LRS_length("tomorrow"))
  print(find_LRS_length("aabdbcec"))
  print(find_LRS_length("fmff"))


main()